pkg/infra/main_helper.go

```diff
@@ -155,13 +155,85 @@ func damageCalculation(fightRoundResult decision.FightResult) {
 
 func updateInternalStates(immutableFightRounds *commons.ImmutableList[decision.ImmutableFightResult], votesResult *immutable.Map[decision.Intent, uint]) {
 	var wg sync.WaitGroup
-	for _, a := range agentMap {
+	for id, a := range agentMap {
+		id := id
 		a := a
 		wg.Add(1)
 		go func(wait *sync.WaitGroup) {
-			a.Strategy.UpdateInternalState(immutableFightRounds, votesResult)
+			a.HandleUpdateInternalState(globalState.AgentState[id], immutableFightRounds, votesResult)
 			wait.Done()
 		}(&wg)
 	}
 	wg.Wait()
 }
+
+/*
+	Hp Pool Helpers
+*/
+
+type HpPoolDonation struct {
+	agentId  commons.ID
+	donation uint
+}
+
+func checkHpPool() {
+	if globalState.HpPool >= globalState.MonsterAttack {
+		logging.Log(logging.Info, logging.LogField{
+			"Original HP Pool":  globalState.HpPool,
+			"Monster Damage":    globalState.MonsterAttack,
+			"HP Pool Remaining": globalState.HpPool - globalState.MonsterAttack,
+		}, fmt.Sprintf("Skipping level %d through HP Pool", globalState.CurrentLevel))
+
+		globalState.HpPool -= globalState.MonsterAttack
+		globalState.MonsterHealth = 0
+	}
+}
+
+func updateHpPool() {
+	var wg sync.WaitGroup
+	donationChan := make(chan HpPoolDonation, len(agentMap))
+	for id, a := range agentMap {
+		// fmt.Print(a.BaseAgent.AgentState().Hp)
+		id := id
+		a := a
+		wg.Add(1)
+		go func(wait *sync.WaitGroup, donationChan chan HpPoolDonation) {
+			donation := a.HandleDonateToHpPool(globalState.AgentState[id])
+			donationChan <- HpPoolDonation{id, donation}
+			wait.Done()
+		}(&wg, donationChan)
+	}
+	wg.Wait()
+
+	sum := uint(0)
+	for i := 0; i < len(agentMap); i++ {
+		agentDonation := <-donationChan
+
+		agentHp := globalState.AgentState[agentDonation.agentId].Hp
+		if agentDonation.donation > agentHp {
```

Comment:
Very minor but should this be `>=` instead of `>`

Reply:
I don’t think it should matter, since we set the donation amount to above their health in the case it’s over.

Reply:
I was thinking if an agent has 10 Hp, and they donate 10 Hp, then they would be able to continue performing actions with 0 health

Reply:
Yeah — exactly, it should be `>=` or else agents wouldn’t be purged with sacrifice.